Suggested "fix" if intermediate states leave the constraint hilbert space #1973

dsmic opened the pull request:
Even if the operator you defined may not leave the constraint hilbert space, it is internaly expanded into a sum, which might leave the constraint hilbert space.

I suggest a fix, which is only triggered by failing numbers = hilb.states_to_numbers(x_prime).

It will for sure need some discussions I think, but it might be of interest?!

Detlef

gcarleo commented:
I am not sure this would scale beyond small Hilbert spaces ? we just use a function that checks if the constraint is satisfied or not (there should be one already, if I'm not mistaken !)

gcarleo commented:
ideally we want to define the operator 'in' such that something like 'check= conf in hiilbert' is doable

dsmic commented:
> ideally we want to define the operator 'in' such that something like 'check= conf in hiilbert' is doable

yes, if you define it that way it is transformed into a sum, where each summand leaves the constraint hilbert space, but the sum would not

gcarleo commented:
what I am saying is that states_to_numvera works only if the Hilbert space Is indexable, so it's not a general solution for constrained spaces! in amy case, in my view the fact that the operator returns something outside of its Hilbert space Is a problem with the operator, that should be addressed in there. in other words I think it's wrong to return states that are not in the Hilbert, even though eventually their matrix elements are zero when summing up

PhilipVinc commented:
> ideally we want to define the operator 'in' such that something like 'check= conf in hiilbert' is doable

First we'd need to define this for bit strings in the basis...
It would require defining `__contains__` on hilbert spaces. 
Though this is not so strictly needed now...
For operators it's trickier

> what I am saying is that states_to_numvera works only if the Hilbert space Is indexable

The PR is to fix `to_sparse` so it naturally will only work for indexable spaces. 
Note that while the default definition of indexable requires that the underlying unconstrained Hilbert space is itself indexable, we support specialised versions of `is_indexable` that work natively in the constrained space. Similar to what QuSpin does.

> in other words I think it's wrong to return states that are not in the Hilbert, even though eventually their matrix elements are zero when summing up

Yeah, but if we can fix the conversion to `to_dense` once and for all, it would be acceptable for me. Fixing the operators is a much harder problem, and I'm unsure we could have a general fix that is performant.

@dsmic can you give me an example operator that would have failed to convert to dense without your PR, but that can successfully be converted with your patch?

Note that I suspect your implementation is checking that the total sum of the invalid mess (across all rows) adds up to 0, but you should check the condition row-by-row...

A simpler implementation I think could be something along those lines (note that it would only work for Homogeneous Hilbert...

```python
def to_sparse(self) -> _csr_matrix:
    #....
    sections1[0] = 0

    if not hilb.constrained:
        numbers = hilb.states_to_numbers(x_prime)
    else:
        # check all x_prime are in the hilbert space
        # invalid_configurations = x_prime in hilb
        # below only works for HomogeneousHilbert and childs, but
        # could be addressed...
        valid_configurations = hilb.constraint(x_prime)
        
        if all(valid_configurations):
            numbers = hilb.states_to_numbers(x_prime)
        else:
            unique_invalid_configs = np.unique(x_prime[~valid_configurations])
            for i in sections:
                # do some magic
                # check that they add up correctly to 0
                # raise error if not
            # assume from here that invalid configs add up to 0
            removed_indices = np.where(~valid_configurations)[0]

            # Apply the mask to all arrays at once
            x_prime = np.delete(x_prime, removed_indices, axis=0)
            mels = np.delete(mels, removed_indices)

            # Adjust sections1 based on removed indices
            adjustment = np.searchsorted(removed_indices, sections1, side='left')
            sections1 -= adjustment
            numbers = hilb.states_to_numbers(x_prime)
    return car_matrix(...)
```

PhilipVinc commented:
FYI, if we can get to a code that is correct and you contribute a test case, I'd be happy to get this merged. It will take a few iterations though. For example, I'd rather not have a try-catch statement around those things to avoid swallowing other kind of exceptions that might be important.

gcarleo commented:
yes I understand, but if you fix the operator (why it's hard ? you need to check in get_conn that the returned elements are valid, or remove them) you fix everything,.not only to_sparse

gcarleo commented:
also I don't understand why we need to this with state_to_number, this seems a bit dirty to me. at some point there was an explicit function to check if the constraint was satisfied, and it was used to trim off states. has it been removed ?

PhilipVinc commented:
> yes I understand, but if you fix the operator (why it's hard ? you need to check in get_conn that the returned elements are valid, or remove them) you fix everything,.not only to_sparse

Well, you don't want to do the checking in `get_conn` if you know you don't need it. It can get costly if you have tons of connected entries. 
For converting `to_sparse` I'm fine because anyhow this has an exponential, high cost and you already would be doing it only once, not in an hot loop.

I think it's fixable, but it raises a few questions of how you can specify that you don't want the checking to be done. Also for jax we'd need to find a way to precompute the actual number of connected entries if we remove some...

> also I don't understand why we need to this with state_to_number, this seems a bit dirty to me. 

You introduced it! https://github.com/netket/netket/blob/30a2a931356e5e9093607648aa5f42e3c50c5b80/NetKet/Hilbert/hilbert_index.hpp#L70
I mean, how do we convert the state (a bit string) to an integer for filling the sparse matrix?

> `hilbert.constraint(states)` exists for that reason. Trimming must be done by hand.

at some point there was an explicit function to check if the constraint was satisfied, and it was used to trim off states. has it been removed ?

dsmic commented:
@PhilipVinc  
The operator I use is the projection operator:
```
if enable_projection:
    from netket.experimental.operator.fermion import destroy as c_o
    from netket.experimental.operator.fermion import create as cdag_o
    from netket.experimental.operator.fermion import number as nc

    def c(hi, site, spin):
        return c_o(hi, site, spin) * (1 - nc(hi, site, -spin))

    def cdag(hi, site, spin):
        return cdag_o(hi, site, spin) * (1 - nc(hi, site, -spin))
else:
    from netket.experimental.operator.fermion import destroy as c
    from netket.experimental.operator.fermion import create as cdag
    from netket.experimental.operator.fermion import number as nc
```

The try catch was my way to avoid the overhead (which was quite significant without the optimizations), if the constraint Hilbert space was not left. 

"Note that I suspect your implementation is checking that the total sum of the invalid mels (across all rows) adds up to 0, but you should check the condition row-by-row..."

I am checking the sum of mels, which belong to one x_prime, so I do it row-by-row?!
```
for i in range(x_prime.shape[0]):
                key = tuple(x_prime[i])
                if key in x_prime_dict:
                    x_prime_dict[key] += mels[i]
                else:
                    x_prime_dict[key] = mels[i]
```

@ gcarleo 
yes you are right. I think it is a little dirty, and there might be an earlier place to fix it, but as far as I understand it is a conceptional problem in constraint Hilbert space (I am saying this with only some weeks into this subject, so please ignore me, if I talk nonsense):

If intermediate results are not allowed to leave the constraint Hilbert space you break a lot of calculation rules e.g.:
cdag_o(hi, site, spin) * (1 - nc(hi, site, -spin)) == cdag_o(hi, site, spin) * 1 - cdag_o(hi, site, spin) * nc(hi, site, -spin))

would not be a valid transformation.

Additionally you break anyway, that it does not matter if you apply the operator to the left or to the right:
<psi|op|psi>

if op is destruction operator, it never leaves the hilbert space applied to the right, but can leave it if applied to the left.

PhilipVinc commented:
Can you give me some code that actually builds the operator?

dsmic commented:
Sure:

```
from numba import jit, config
# config.DISABLE_JIT = True


# checking tranlational symmetry of the Hamiltonian
do_experimental_constraint = True
enable_projection = True

L1=4
L2=4
n_elecs=4

do_symmetrize_operators = True

up, down = +1, -1


if enable_projection:
    from netket.experimental.operator.fermion import destroy as c_o
    from netket.experimental.operator.fermion import create as cdag_o
    from netket.experimental.operator.fermion import number as nc


    def c(hi, site, spin):
        """
        This was, what I thought was necessary, but c_o is violating?
        """
        return c_o(hi, site, spin) * (1 - nc(hi, site, -spin))

    def cdag(hi, site, spin):
        """
        This was, what I thought was necessary, but c_o is violating?
        """
        return cdag_o(hi, site, spin) * (1 - nc(hi, site, -spin))
else:
    from netket.experimental.operator.fermion import destroy as c
    from netket.experimental.operator.fermion import create as cdag
    from netket.experimental.operator.fermion import number as nc

import netket as nk
from netket import experimental as nkx
import numpy as np
import matplotlib.pyplot as plt
import jax.numpy as jnp

N_sites = L1*L2
N_up    = (n_elecs+1)//2
N_dn    = n_elecs//2

g = nk.graph.Grid(extent=(L1, L2), pbc=True)  # 2D square lattice with periodic boundary conditions

class AvoidDoubleOccupancy(nk.hilbert.constraint.DiscreteHilbertConstraint):
    def __call__(self, x):
        # Reshape to (batch, 2, L1 * L2)
        # print(x.shape)
        x = x.reshape(-1, 2, N_sites)
        # jax.debug.print("x -- {x}",x=x)
        # Sum over spin dimension to get total occupancy per site
        x_sum = jnp.sum(x, axis=1)  # Shape: (batch, L1*L2)
        
        # Check all occupancies <= 1
        valid = jnp.all(x_sum <= 1, axis=-1)  # Shape: (batch,)
        
        # other constraints do this, but it is not documented
        if valid.shape[0] == 1:
            return valid[0]
        return valid.reshape(-1)

    def __hash__(self):
        return hash(("AvoidDoubleOccupancy",))

    def __eq__(self, other):
        return isinstance(other, AvoidDoubleOccupancy)
    

if do_experimental_constraint:
    constraint = AvoidDoubleOccupancy()
else:
    constraint = None
hi = nkx.hilbert.SpinOrbitalFermions(N_sites, 
                                     s = 1/2, 
                                     n_fermions_per_spin = (N_up, N_dn), 
                                     constraint=constraint
                                     )


print(f"hi.size = {hi.size}")
print(f"hi.n_states = {hi.n_states}")


# Define the hopping parameter (t) and exchange interaction (J)
t = 3.0  # Hopping amplitude
J = Jz = Jp = 1.0  # Exchange interaction strength

# ------------- define Hamiltonian ------------------------
def Sz(site):
  return 1/2*(nc(hi, site, up) - nc(hi, site, down))

def Splus(site):
  return cdag(hi, site,up)*c(hi, site,down)

def Sminus(site):
  return cdag(hi, site,down)*c(hi, site,up)

up, down = +1, -1
ha = 0.0
for sz in (up, down):
    for u, v in g.edges():
        ha += -t*(cdag(hi, u, sz) * c(hi, v, sz)  + cdag(hi, v, sz) * c(hi, u, sz))

for u,v in g.edges():
    ha += Jz*Sz(u)*Sz(v)
    ha += 1/2*Jp*Splus(u)*Sminus(v)
    ha += 1/2*Jp*Sminus(u)*Splus(v)
    ha -= 1/4*Jz*(nc(hi,u,up) + nc(hi,u,down))*(nc(hi,v,up) + nc(hi,v,down))


num_eigenstate = 0
# Exact Diagonalization
eigenvalues, eigenvectors = nk.exact.lanczos_ed(ha, k = 10, compute_eigenvectors=True)  # Compute eigenvalues and eigenvectors
exact_ground_energy = eigenvalues[num_eigenstate]
ground_state = eigenvectors[:, num_eigenstate]  # Ground state wavefunction

print(f"Edges: {g.edges()}")
print(f"Exact State Energy for {num_eigenstate} state: {exact_ground_energy}")

def two_correlation(hilbert, site1, site2):
    """
    Creates a two-point correlation operator between two sites:
    O(site1, site2) = c†(site1) c(site2)
    """
    op = 0.0
    op += cdag(hilbert, site1, up) * c(hilbert, site2, up) 
    op += cdag(hilbert, site1, down) * c(hilbert, site2, down) 
    return op

def opposite_spin_correlation(hilbert, site1, site2):
    """
    Creates a two-point correlation operator between two sites with opposite spins:
    O(site1, site2) = c†(site1, ↑) c(site2, ↓)
    """
    op = 0.0
    op += nc(hilbert, site1, up) * nc(hilbert, site2, down)
    op += nc(hilbert, site1, down) * nc(hilbert, site2, up)
    # op += cdag(hilbert, site1, up) *cdag(hilbert, site2, down) * c(hilbert, site1, down) * c(hilbert, site2, up)
    return op



site1=0
print(f"spin up at site {site1} {np.vdot(ground_state, nc(hi, site1, up) @ ground_state)}")
site1=1
print(f"spin up at site {site1} {np.vdot(ground_state, nc(hi, site1, up) @ ground_state)}")
site1=0
print(f"spin down at site {site1} {np.vdot(ground_state, nc(hi, site1, down) @ ground_state)}")
site1=1
print(f"spin down at site {site1} {np.vdot(ground_state, nc(hi, site1, down) @ ground_state)}")

site1=0; site2=1
print(f"two_correlation site({site1}) site({site2}) {np.vdot(ground_state, two_correlation(ha.hilbert, site1, site2) @ ground_state)}")
site1=1; site2=2
print(f"two_correlation site({site1}) site({site2}) {np.vdot(ground_state, two_correlation(ha.hilbert, site1, site2) @ ground_state)}")

site1=0; site2=1
print(f"opposite_spin_correlation site({site1}) site({site2}) {np.vdot(ground_state, opposite_spin_correlation(ha.hilbert, site1, site2) @ ground_state)}")
site1=1; site2=2
print(f"opposite_spin_correlation site({site1}) site({site2}) {np.vdot(ground_state, opposite_spin_correlation(ha.hilbert, site1, site2) @ ground_state)}")

```

PhilipVinc commented:
> I am checking the sum of mels, which belong to one x_prime, so I do it row-by-row?!

yeah. By checking the sum of mels corresponding to one x_prime, you are checking that the corresponding column sums to 0.
However what we want to check is that the individual entries are 0?

dsmic commented:
"However what we want to check is that the individual entries are 0?"

Yeah, you might be correct, I will have to think about. But maybe the discarding all states, which leave the constraint hilbert space is the better idea anyway?! And it can be good documented, as you just have to say, we project it on the constraint hilbert space...

something like (which is not performant, one has to use hash I think, but runs with the same result as before):

```
    def get_conn_flattened(self, x, sections, pad=False):
        r"""Finds the connected elements of the Operator.

        Starting from a given quantum number x, it finds all other quantum numbers x' such
        that the matrix element :math:`O(x,x')` is different from zero. In general there
        will be several different connected states x' satisfying this
        condition, and they are denoted here :math:`x'(k)`, for :math:`k=0,1...N_{\mathrm{connected}}`.

        This is a batched version, where x is a matrix of shape (batch_size,hilbert.size).

        Args:
            x: A matrix of shape (batch_size,hilbert.size) containing
                the batch of quantum numbers x.
            sections: An array of size (batch_size) useful to unflatten
                the output of this function.
                See numpy.split for the meaning of sections.

        Returns:
            matrix: The connected states x', flattened together in a single matrix.
            array: An array containing the matrix elements :math:`O(x,x')` associated to each x'.
        """
        self._setup()

        x = concrete_or_error(
            np.asarray,
            x,
            NumbaOperatorGetConnDuringTracingError,
            self,
        )

        assert (
            x.shape[-1] == self.hilbert.size
        ), "size of hilbert space does not match size of x"
        return self._flattened_kernel(
            x,
            sections,
            self.max_conn_size,
            self._orb_idxs,
            self._daggers,
            self._numba_weights,
            self._diag_idxs,
            self._off_diag_idxs,
            self._term_split_idxs,
            self._cutoff,
            pad,
        )

    @staticmethod
    @numba.jit(nopython=True)
    def _flattened_kernel(  # pragma: no cover
        x,
        sections,
        max_conn,
        orb_idxs,
        daggers,
        weights,
        diag_idxs,
        off_diag_idxs,
        term_split_idxs,
        cutoff,
        pad=False,
    ):
        x_prime = np.empty((x.shape[0] * max_conn, x.shape[1]), dtype=x.dtype)
        mels = np.zeros((x.shape[0] * max_conn), dtype=weights.dtype)

        # do not split at the last one (gives empty array)
        term_split_idxs = term_split_idxs[:-1]
        orb_idxs_list = np.split(orb_idxs, term_split_idxs)
        daggers_list = np.split(daggers, term_split_idxs)

        # loop over the batch dimension
        n_c = 0
        for b in range(x.shape[0]):
            xb = x[b, :]

            # we can already fill up with default values
            if pad:
                x_prime[b * max_conn : (b + 1) * max_conn, :] = np.copy(xb)

            non_zero_diag = False
            # first do the diagonal terms, they all generate just 1 term
            for term_idx in diag_idxs:
                mel = weights[term_idx]
                xt = np.copy(xb)
                has_xp = True
                for orb_idx, dagger in zip(
                    orb_idxs_list[term_idx], daggers_list[term_idx]
                ):
                    _, mel, op_has_xp = _apply_operator(
                        xt, orb_idx, dagger, mel, cutoff
                    )
                    if not op_has_xp:
                        has_xp = False
                        continue
                if has_xp and np.any(np.all(x == xb, axis=1)):  # this ignores states, which are not in the hilbert space      <-- This is changed
                    x_prime[n_c, :] = np.copy(xb)  # should be untouched
                    mels[n_c] += mel

                non_zero_diag = non_zero_diag or has_xp

            # end of the diagonal terms
            if non_zero_diag:
                n_c += 1

            # now do the off-diagonal terms
            for term_idx in off_diag_idxs:
                mel = weights[term_idx]
                xt = np.copy(xb)
                has_xp = True
                for orb_idx, dagger in zip(
                    orb_idxs_list[term_idx], daggers_list[term_idx]
                ):
                    xt, mel, op_has_xp = _apply_operator(
                        xt, orb_idx, dagger, mel, cutoff
                    )
                    if not op_has_xp:  # detect zeros
                        has_xp = False
                        continue
                if has_xp and np.any(np.all(x == xt, axis=1)): # this ignores states, which are not in the hilbert space       <-- This is changed
                    x_prime[n_c, :] = np.copy(xt)  # should be different
                    mels[n_c] += mel
                    n_c += 1

            # end of this sample
            if pad:
                n_c = (b + 1) * max_conn

            sections[b] = n_c

        if pad:
            return x_prime, mels
        else:
            return x_prime[:n_c], mels[:n_c]
```

PhilipVinc commented:
> Yeah, you might be correct, I will have to think about. But maybe the discarding all states, which leave the constraint hilbert space is the better idea anyway?! And it can be good documented, as you just have to say, we project it on the constraint hilbert space.

It should be the same cost. You just need to transverse every row once...

PhilipVinc commented:
Moreover this would be 'correct', while what you are doing leads to something that is not identical between its action and matrix representation..

dsmic commented:
OK, I will close this, as it is not correct, and discarding states out of the constraint Hilbert space seems to be the better option.

I will open a new one with a discarding solution for the FermionOperator2nd operator and hopefully be able to link to the discussion of this closed pull request.
